### PR TITLE
Support certify for unseal in gotpm cmd

### PR DESCRIPTION
**Pull request by lulumel0n**
Update gotpm/cmd to support certify unseal.

### CODE REVIEW

File: gotpm/cmd/seal_test.go [hunk 1]
```diff
@@ -104,17 +121,26 @@ func TestUnsealFail(t *testing.T) {
...
 			}
+			unsealArgs := []string{"unseal", "--quiet", "--input", sealedFile, "--output", secretFile}
+			if test.certifyPCRs != "" {
+				unsealArgs = append(unsealArgs, "--certify-pcrs", test.certifyPCRs)
+			}
 
-			RootCmd.SetArgs([]string{"unseal", "--quiet", "--input", sealedFile, "--output", secretFile})
-			if RootCmd.Execute() == nil {
+			RootCmd.SetArgs(unsealArgs)
+			if err := RootCmd.Execute(); err == nil {
```
Comment (stevenrutherford) [inline gotpm/cmd/seal_test.go]: Why do you need to add err, if you aren't going ot use it?

File: gotpm/cmd/seal.go [hunk 2]
```diff
@@ -108,7 +108,18 @@ Thus, algorithm and PCR options are not needed for the unseal command.`,
...
-		secret, err := srk.Unseal(&sealed, nil)
+
+		certifySel, err := getSelection()
+		if err != nil {
+			return err
+		}
+
+		var secret []byte
+		if len(certifySel.PCRs) > 0 {
+			secret, err = srk.Unseal(&sealed, tpm2tools.CertifyCurrent{PCRSelection: certifySel})
```
Comment (stevenrutherford) [inline gotpm/cmd/seal.go]: This is fine as is, but I probably would have written this slightly differently (and I would have done the same in the prior PR):
```
cOpt := tpm2tools.CertifyCurrent{PCRSelection: certifySel}
if len(certifySel.PCRs) == 0 {
  cOpt = nil
}
secret, err = srk.Unseal(&sealed, cOpt)
```

Not a mandatory change, I just feel like it removes some duplication.

File: gotpm/cmd/seal_test.go [hunk 3]
```diff
@@ -35,13 +35,19 @@ func TestSealPlain(t *testing.T) {
...
+		algo        string
+		sealPCRs    string
+		certifyPCRs string
 	}{
-		{"RSA", "rsa", ""},
-		{"RSAwithPCRs", "rsa", "7,8"},
-		{"ECCwithPCRs", "ecc", "7"},
+		{"RSA", "rsa", "", ""},
+		{"ECC", "ecc", "", ""},
+		{"RSAWithSealPCR", "rsa", "7", ""},
```
Comment (stevenrutherford) [inline gotpm/cmd/seal_test.go]: Nit: I'd probably call this RSASeal (In general: RSAVerbOptionalverb).

File: gotpm/cmd/seal_test.go [hunk 4]
```diff
@@ -84,17 +95,23 @@ func TestUnsealFail(t *testing.T) {
...
+		tpmFunc          func(io.ReadWriter, int) error
+		tpmFuncParameter int
 	}{
 		// TODO(joerichey): Add test that TPM2_Reset make unsealing fail
-		{"ExtendPCR", func(rw io.ReadWriter) error {
-			pcrToExtend := tpmutil.Handle(23)
-			extension := bytes.Repeat([]byte{0xAA}, sha256.Size)
-			return tpm2.PCRExtend(rw, pcrToExtend, tpm2.AlgSHA256, extension, "")
-		}},
+		{"ExtendSealPCR", "23", "", extendPCR, 23},
```
Comment (stevenrutherford) [inline gotpm/cmd/seal_test.go]: The PCR is extended, so call it ExtendPCRAndSeal or something like that.

File: gotpm/cmd/seal_test.go [hunk 5]
```diff
@@ -84,17 +95,23 @@ func TestUnsealFail(t *testing.T) {
...
+		return tpm2.PCRExtend(rw, pcrToExtend, tpm2.AlgSHA256, extension, "")
+	}
 
 	tests := []struct {
-		name    string
-		tpmFunc func(io.ReadWriter) error
+		name             string
+		sealPCRs         string
+		certifyPCRs      string
+		tpmFunc          func(io.ReadWriter, int) error
```
Comment (stevenrutherford) [inline gotpm/cmd/seal_test.go]: Something feels off about expanding this with an additional parameter.

I would remove the tpmFunc parameter, and just make a helper for extendPCR. If Joe (or someone else) wants to make a parameter so we can call a tpmFunc in between seal/unseal to resolve that TODO, they can add back the tpmFunc parameter.

File: gotpm/cmd/seal_test.go (hunk 1 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: Removed. Used when debugging, forget to clean.

File: gotpm/cmd/seal_test.go (hunk 3 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: RSASeal
RSASealWithPCR
RSACertifyWithPCR
RSASealAndCertifyWithPCR

File: gotpm/cmd/seal_test.go (hunk 4 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: ExtendPCRAndUnseal
ExtendPCRAndCertify
ExtendPCRAndCertify2

File: gotpm/cmd/seal_test.go (hunk 5 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: The problem is we might not need `extendPCR` in every test cases in the future. We need a way to execute some (0 to n) functions between seal/unseal.

File: gotpm/cmd/seal.go (hunk 2 above)
Comment (lulumel0n) [reply gotpm/cmd/seal.go]: done

File: gotpm/cmd/utils.go [hunk 6]
```diff
@@ -47,18 +47,30 @@ func addIndexFlag(cmd *cobra.Command) {
...
+func addCertifyPCRFlag(cmd *cobra.Command) {
+	cmd.PersistentFlags().IntSliceVar(&pcrs, "certify-pcrs", nil,
+		"comma separated list of certify PCR numbers (optional, currently only support sha256 PCR hash)")
+}
+
+// initPCRs will initialize the pcr var. This function is only useful in some test cases, where
+// there are consecutive Execute() calls. But variables like pcrs here are not initialize for every
+// Execute() calls, we need to manually "flush" pcrs value in last Execute() cmd.
+func initPCRs() {
+	pcrs = []int{}
```
Comment (josephlr) [inline gotpm/cmd/utils.go]: Delete this, if we only need it for tests, just write the code in the test.

File: gotpm/cmd/utils.go [hunk 7]
```diff
@@ -47,18 +47,30 @@ func addIndexFlag(cmd *cobra.Command) {
...
 func addHashAlgoFlag(cmd *cobra.Command) {
 	cmd.PersistentFlags().StringVar(&hashAlgo, "hash-algo", "sha256",
-		"Hash Algorithm, \"sha1\",  \"sha256\", or \"sha384\"")
+		"hash algorithm, \"sha1\",  \"sha256\", or \"sha384\"")
+}
+
+func addCertifyPCRFlag(cmd *cobra.Command) {
+	cmd.PersistentFlags().IntSliceVar(&pcrs, "certify-pcrs", nil,
+		"comma separated list of certify PCR numbers (optional, currently only support sha256 PCR hash)")
+}
```
Comment (josephlr) [inline gotpm/cmd/utils.go]: We don't need another flag here, just use the existing `--pcrs` flag.

File: gotpm/cmd/seal_test.go [hunk 8]
```diff
@@ -85,16 +97,29 @@ func TestUnsealFail(t *testing.T) {
...
+		return tpm2.PCRExtend(rw, pcrToExtend, tpm2.AlgSHA256, extension, "")
+	}
+
 	tests := []struct {
-		name    string
-		tpmFunc func(io.ReadWriter) error
+		name              string
+		sealPCRs          string
+		certifyPCRs       string
+		tpmFuncs          []intermediateFunc
```
Comment (josephlr) [inline gotpm/cmd/seal_test.go]: If we're always have the exact same code for `tpmFuncs` just move the `intermediteFunc` code into the test for loop, and remove this indirection.

File: gotpm/cmd/seal_test.go [hunk 9]
```diff
@@ -104,16 +129,27 @@ func TestUnsealFail(t *testing.T) {
...
+			initPCRs()
 
-			if err := test.tpmFunc(rwc); err != nil {
-				t.Fatal(err)
+			for i, f := range test.tpmFuncs {
+				if err := f(rwc, test.tpmFuncParameters[i]); err != nil {
+					t.Fatal(err)
+				}
+			}
+			unsealArgs := []string{"unseal", "--quiet", "--input", sealedFile, "--output", secretFile}
```
Comment (josephlr) [inline gotpm/cmd/seal_test.go]: Add empty line above this line

File: gotpm/cmd/seal_test.go [hunk 10]
```diff
@@ -104,16 +129,27 @@ func TestUnsealFail(t *testing.T) {
...
+			if test.certifyPCRs != "" {
+				unsealArgs = append(unsealArgs, "--certify-pcrs", test.certifyPCRs)
 			}
 
-			RootCmd.SetArgs([]string{"unseal", "--quiet", "--input", sealedFile, "--output", secretFile})
+			RootCmd.SetArgs(unsealArgs)
```
Comment (josephlr) [inline gotpm/cmd/seal_test.go]: Delete the empty line above this line

File: gotpm/cmd/seal_test.go [hunk 11]
```diff
@@ -104,16 +129,27 @@ func TestUnsealFail(t *testing.T) {
...
-			RootCmd.SetArgs([]string{"seal", "--quiet", "--input", secretFile, "--output", sealedFile, "--pcrs", "23"})
+			sealArgs := []string{"seal", "--quiet", "--input", secretFile, "--output", sealedFile}
+			if test.sealPCRs != "" {
+				sealArgs = append(sealArgs, "--pcrs", test.sealPCRs)
+			}
+			RootCmd.SetArgs(sealArgs)
 			if err := RootCmd.Execute(); err != nil {
 				t.Error(err)
 			}
+			initPCRs()
```
Comment (josephlr) [inline gotpm/cmd/seal_test.go]: just manually set pcrs here (and add a note explaining why)

File: gotpm/cmd/seal.go [hunk 12]
```diff
@@ -108,7 +107,16 @@ Thus, algorithm and PCR options are not needed for the unseal command.`,
 		defer srk.Close()
 
 		fmt.Fprintln(debugOutput(), "Unsealing data")
-		secret, err := srk.Unseal(&sealed, nil)
+
```
Comment (josephlr) [inline gotpm/cmd/seal.go]: We need to change the description above to note what the (optional) `--pcrs` flag does.

File: gotpm/cmd/seal.go [hunk 13]
```diff
@@ -132,5 +140,6 @@ func init() {
 	// PCRs and hash algorithm only used for sealing
 	addPCRsFlag(sealCmd)
 	addHashAlgoFlag(sealCmd)
+	addCertifyPCRFlag(unsealCmd)
```
Comment (josephlr) [inline gotpm/cmd/seal.go]: Just use `addPCRsFlag`

File: gotpm/cmd/utils.go (hunk 6 above)
Comment (lulumel0n) [reply gotpm/cmd/utils.go]: done

File: gotpm/cmd/utils.go (hunk 7 above)
Comment (lulumel0n) [reply gotpm/cmd/utils.go]: Done. I feel it can have more descriptive help message if using a separate flag..

File: gotpm/cmd/seal_test.go (hunk 11 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: done

File: gotpm/cmd/seal.go (hunk 12 above)
Comment (lulumel0n) [reply gotpm/cmd/seal.go]: `The optional --pcrs flag allows the user to certify the blob that has been sealed
under the same PCR values as current with the given PCR numbers.`

File: gotpm/cmd/seal_test.go (hunk 8 above)
Comment (lulumel0n) [reply gotpm/cmd/seal_test.go]: I can do that. But in the future we probably need more "intermediate" functions other than extendPCR. For now I can move extendPCR into the loop for simplicity.

File: tpm2tools/keys.go [hunk 14]
```diff
@@ -278,6 +278,15 @@ func (k *Key) Unseal(in *proto.SealedBytes, cOpt CertifyOpt) ([]byte, error) {
 		if _, _, err = tpm2.CertifyCreation(k.rw, "", sealed, tpm2.HandleNull, nil, creationHash.Sum(nil), tpm2.SigScheme{}, ticket); err != nil {
 			return nil, fmt.Errorf("failed to certify creation: %v", err)
 		}
+		// verify certify PCRs haven't been modified
+		decodedCreationData, err := tpm2.DecodeCreationData(in.GetCreationData())
+		if err != nil {
+			return nil, fmt.Errorf("failed to decode creation data: %v", err)
+		}
+		if subtle.ConstantTimeCompare(decodedCreationData.PCRDigest, computePCRDigest(in.GetCertifiedPcrs())) == 0 {
```
Comment (josephlr) [inline tpm2tools/keys.go]: Before checking the equality of digests, we need to make sure the selections match.

File: gotpm/cmd/seal.go [hunk 15]
```diff
@@ -132,5 +144,6 @@ func init() {
 	// PCRs and hash algorithm only used for sealing
 	addPCRsFlag(sealCmd)
 	addHashAlgoFlag(sealCmd)
+	addPCRsFlag(unsealCmd)
```
Comment (josephlr) [inline gotpm/cmd/seal.go]: We also need to add the HashAlgoFlag here for unseal, as we want the ability to select a specific hash bank.

File: gotpm/cmd/seal.go [hunk 16]
```diff
@@ -81,7 +80,11 @@ using the TPM. This operation will fail if used on a different TPM, or if the
 Platform Control Registers (PCRs) are in the incorrect state.
 
 All the necessary data to decrypt the sealed input is present in the input blob.
-Thus, algorithm and PCR options are not needed for the unseal command.`,
+Thus, algorithm and seal PCR options are not needed for the unseal command.
+
+The optional --pcrs flag allows the user to certify the blob that has been sealed
+under the same PCR values as current with the given PCR numbers.
+`,
```
Comment (josephlr) [inline gotpm/cmd/seal.go]:
```
All the necessary data to decrypt the sealed input is present in the input blob.
We do not need to specify the PCRs used for unsealing.

We do support an optional "certification" process. A list of PCRs may be
provided with --pcrs, and the unwrapping will fail if the PCR values when
sealing differ from the current PCR values. This allows for verification of the
machine state when sealing took place.
```

File: tpm2tools/keys.go (hunk 14 above)
Comment (lulumel0n) [reply tpm2tools/keys.go]: add a function in PCR: EqualsPCRSelections()

File: gotpm/cmd/seal.go (hunk 16 above)
Comment (lulumel0n) [reply gotpm/cmd/seal.go]: done

File: gotpm/cmd/seal.go (hunk 15 above)
Comment (lulumel0n) [reply gotpm/cmd/seal.go]: No longer needed, as we will use the CertifyHashAlgTpm

**Pull request merged**